plugin-server/src/main/ingestion-queues/session-recording/blob-ingester/session-manager.ts

```diff
@@ -230,82 +238,65 @@ export class SessionManager {
-
-                            fileStream.on('error', (err) => {
-                                // TODO: What should we do here?
-                                status.error('🧨', 'blob_ingester_session_manager readstream errored', {
-                                    ...this.logContext(),
-                                    error: err,
-                                })
-
-                                this.captureException(err)
-                            })
-
-                            this.inProgressUpload = new Upload({
-                                client: this.s3Client,
-                                params: {
-                                    Bucket: this.serverConfig.OBJECT_STORAGE_BUCKET,
-                                    Key: dataKey,
-                                    Body: fileStream,
-                                },
-                            })
-
-                            await this.inProgressUpload.done()
-                            fileStream.close()
-                            clearTimeout(flushTimeout)
-                            resolve()
-                        } catch (error) {
-                            reject(error)
-                        }
-                    })
-                } catch (error) {
-                    reject(error)
-                }
+            if (count === 0) {
+                throw new Error("Can't flush empty buffer")
+            }
+
+            if (!eventsRange) {
+                throw new Error("Can't flush buffer due to missing eventRange")
+            }
+
+            const { firstTimestamp, lastTimestamp } = eventsRange
+            const baseKey = `${this.serverConfig.SESSION_RECORDING_REMOTE_FOLDER}/team_id/${this.teamId}/session_id/${this.sessionId}`
+            const timeRange = `${firstTimestamp}-${lastTimestamp}`
+            const dataKey = `${baseKey}/data/${timeRange}`
+
+            // We want to ensure the writeStream has ended before we read from it
+            await new Promise((r) => fileStream.end(r))
+
+            const readStream = createReadStream(file, 'utf-8')
+            const gzippedStream = readStream.pipe(zlib.createGzip())
+
+            readStream.on('error', (err) => {
+                // TODO: What should we do here?
+                status.error('🧨', 'blob_ingester_session_manager readstream errored', {
+                    ...this.logContext(),
+                    error: err,
+                })
+
+                this.captureException(err)
             })
 
+            this.inProgressUpload = new Upload({
+                client: this.s3Client,
+                params: {
+                    Bucket: this.serverConfig.OBJECT_STORAGE_BUCKET,
+                    Key: dataKey,
+                    Body: gzippedStream,
+                },
+            })
+
+            await this.inProgressUpload.done()
+            readStream.close()
```

Comment:
```suggestion
            await new Promise((resolve, reject) => {readStream.close((err) => err ? reject(err) : resolve())})
```

That's a lot of brackets to type into a github search box

readstream close takes a callback which receives err|null. We could wait here too?

interestingly readstream also has a `bytesRead` property which we could use instead of sizeEstimate 🤷

Reply:
I feel like I at some point purposefully did not await that... But I'm not sure why 😅 
I'll take a look at that in a follow up maybe